create nft floor price spell - Ready for review

henrystats opened the pull request:
Brief comments on the purpose of your changes:

**For Dune Engine V2**

I've checked that:

### General checks:
* [ ] I tested the query on dune.com after compiling the model with dbt compile (compiled queries are written to the target directory)
* [ ] I used "refs" to reference other models in this repo and "sources" to reference raw or decoded tables 
* [ ] if adding a new model, I added a test
* [ ] the filename is unique and ends with .sql
* [ ] each sql file is a select statement and has only one view, table or function defined  
* [ ] column names are `lowercase_snake_cased`
* [ ] if adding a new model, I edited the dbt project YAML file with new directory path for both models and seeds (if applicable)
* [ ] if wanting to expose a model in the UI (Dune data explorer), I added a post-hook in the JINJA config to add metadata (blockchains, sector/project, name and contributor Dune usernames)

### Pricing checks:
* [ ] `coin_id` represents the ID of the coin on coinpaprika.com
* [ ] all the coins are active on coinpaprika.com (please remove inactive ones)

### Join logic:
* [ ] if joining to base table (i.e. ethereum transactions or traces), I looked to make it an inner join if possible

### Incremental logic:
* [ ] I used is_incremental & not is_incremental jinja block filters on both base tables and decoded tables
  * [ ] where block_time >= date_trunc("day", now() - interval '1 week')
* [ ] if joining to base table (i.e. ethereum transactions or traces), I applied join condition where block_time >= date_trunc("day", now() - interval '1 week')
* [ ] if joining to prices view, I applied join condition where minute >= date_trunc("day", now() - interval '1 week')

dune-eng commented:
Workflow run id 4078030797 approved.

dune-eng commented:
Workflow run id 4078140639 approved.

dune-eng commented:
Workflow run id 4078140625 approved.

augustog commented on models/nft/ethereum/nft_ethereum_floor_price.sql:
```diff
@@ -0,0 +1,126 @@
+{{ config(
+    alias = 'floor_price',
```
one good point of feedback from @0xRobin: this is a good base for a collection stats table, so I would rename the spell to `collection_stats`

henrystats replied:
ty ser, I'll add tests later tonight and update the naming.

henrystats replied:
tests added and name updated

augustog reviewed (commented):
It's overall looking good! My only concern would be probably adding a few (unit) tests to ensure it behaves as expected.

I made a smol comment re naming, otherwise lgtm

0xRobin commented on models/nft/ethereum/nft_ethereum_floor_price.sql:
```diff
@@ -0,0 +1,126 @@
+    FROM 
+    {{ ref('nft_trades') }} 
+    WHERE blockchain = 'ethereum'
+    AND trade_type = 'Single Item Trade'
```
nft.trades is already setup so each row contains only one `token_id` traded.
It might thus be possible to drop this conditional.

One row might still contain a trade with multiple ERC1155s though, if you don't want to deal with that I suggest using `number_of_items = 1` as filter.

henrystats replied:
okaiee ser, I'll update thank you very much!

henrystats replied:
updated

0xRobin commented on models/nft/ethereum/nft_ethereum_floor_price.sql:
```diff
@@ -0,0 +1,126 @@
+        src.block_date, 
+        src.nft_contract_address,
+        percentile_cont(.05) WITHIN GROUP 
+            (ORDER BY src.amount_original/(
+                CASE 
+                    WHEN src.currency_symbol IN ('ETH', 'WETH') THEN 1 
+                    ELSE prices.price
+                END
```
From my understanding this price calculation doesn't hold up.
You're doing:

For ETH/WETH trades
`ETH amount / 1`

For other ERC20 trades
`ERC20 amount / ETH price`

henrystats replied:
fixed

0xRobin commented on models/nft/ethereum/nft_ethereum_floor_price.sql:
```diff
@@ -0,0 +1,126 @@
+SELECT 
+    d.day as block_date, 
+    d.nft_contract_address, 
+    COALESCE(prof.currency_volume, 0) as currency_volume,
+    COALESCE(prof.trades, 0) as trades, 
+    COALESCE(prof.fifth_percentile, 0) as fifth_percentile,
+    COALESCE(prof.currency_min, 0) as currency_min, 
+    COALESCE(prof.currency_max, 0) as currency_max 
```
given you're not grouping on different currencies used, I think it's good to adapt the naming here to include ETH or USD here instead of  the generic currency so it's clear in what currency the columns are expressed.

henrystats replied:
agreed sir, I'll update now.

henrystats replied:
updated.

0xRobin reviewed (commented):
I gave this a quick scan and left some comments.

dune-eng commented:
Workflow run id 4123348221 approved.

dune-eng commented:
Workflow run id 4123348212 approved.

dune-eng commented:
Workflow run id 4123504730 approved.

dune-eng commented:
Workflow run id 4123504701 approved.

dune-eng commented:
Workflow run id 4123599703 approved.

dune-eng commented:
Workflow run id 4123599733 approved.

dune-eng commented:
Workflow run id 4123684489 approved.

dune-eng commented:
Workflow run id 4123684486 approved.

dune-eng commented:
Workflow run id 4123835969 approved.

dune-eng commented:
Workflow run id 4123835982 approved.

henrystats commented:
@augustog, @0xRobin I've updated the naming & also added some seeds to test some of the results. 


I wrote a simple & separate query for the seed results - https://dune.com/queries/1978740?d=1

I'm now facing some rounding error with the tests/seeds not sure how to write it in way that'd make it work. The seed results are correct just not rounded properly which is weird.

dune-eng commented:
Workflow run id 4123884079 approved.

dune-eng commented:
Workflow run id 4123884101 approved.

0xRobin commented:
@henrystats I'm sorry can I suggest a final renaming? 
Just want to make sure it's as future proof and clear as possible so we don't need to rename something later on when there's already a bunch of queries depending on it.

currency_volume_eth -> volume_eth
fifth_percentile_eth -> price_p5_eth
currency_min_eth -> price_min_eth
currency_max_eth -> price_max_eth

The rounding problems can get a bit annoying, I'm sorry about that. 
Unfortunately I've not found why they seem to be inconsistent sometimes. 
I'd say try some combinations and if it doesn't work drop a row from the seed if necessary.

henrystats commented:
@0xRobin yes sure sir, I'll update now.

I've tried multiple combinations and at least one of it is always failing :(. Lemme try another and see if it works.

dune-eng commented:
Workflow run id 4124727359 approved.

dune-eng commented:
Workflow run id 4124727360 approved.

henrystats commented:
column names updated, tests still failing due to the rounding errors.

jeff-dude commented:
> column names updated, tests still failing due to the rounding errors.

i haven't dove deep into the rounding errors from seeds, but i'm wondering if it makes sense to change the test approach here from seeds to just a query in the `tests/nft/ethereum/` directory? if you have a CTE for `nft.trades` logic with subset of data, CTE for new model logic with same subset, then join together and perform calculation there? ensure it returns no rows?

maybe seeds are more important for decoded projects -> sector spells, where a simpler query could be used here?

henrystats commented:
Hola ser @jeff-dude, sorry I don't really understand...

jeff-dude commented:
okay @henrystats -- of the three linked PRs, i edited this one first. here are some notes:
- incremental models are tricky when using day sequence as driving table, then left joining in `nft.trades` data. if you look at latest code, i'm only using incremental filter on days CTE, otherwise the others pull all history in order to aggregate properly
- removed any seed references and used a test specific to this model. my point above was intended to mean that any spells which are not sector-wide (i.e. `dex.trades`, `nft.trades`, `perpetual.trades`, etc) can use more specific tests for their model, rather than the seed approach. i think this makes it easier to build and test. since you read from `nft.trades`, we can feel confident from prior seed tests that data is correct and can focus on the specific test case for this model

does this look to accomplish what you need from this spell?

henrystats commented:
Yes ser @jeff-dude, thank you very much for your help ser!

jeff-dude commented on seeds/nft/nft_collection_stats_seed.csv:
```diff
@@ -0,0 +1,4 @@
+block_date,nft_contract_address,volume_eth,price_p5_eth,price_max_eth,price_min_eth
+2023-01-31,0xbc4ca0eda7647a8ab7c2061c2e118a18a936f13d,732.3399999999999,66.211,310,66.16
+2023-01-19,0xbc4ca0eda7647a8ab7c2061c2e118a18a936f13d,1249.3517,65.45,136.9,65.25
+2023-01-22,0xbc4ca0eda7647a8ab7c2061c2e118a18a936f13d,1594.1301999999998,67.85109,194.2,67.5
```
i see less of a need to force a seed file here, as the spell is built specifically for one team -- even though it's under the `nft` schema which adds a bit of confusion. as these spells mature (or if they do), then i think forcing the seeds and fixing this is a priority

0xRobin replied:
@jeff-dude not sure why my comment was included in my "approval". I assume it was a comment from last week that stayed in the drafts.. :thinking:

jeff-dude replied:
weird!
